File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/diary/DiaryLiveData.java
package com.websarva.wings.android.zuboradiary.ui.diary;

import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.MutableLiveData;

import com.websarva.wings.android.zuboradiary.data.database.DiaryEntity;
import com.websarva.wings.android.zuboradiary.data.database.DiaryItemTitleSelectionHistoryItemEntity;
import com.websarva.wings.android.zuboradiary.data.diary.Condition;
import com.websarva.wings.android.zuboradiary.data.diary.ItemNumber;
import com.websarva.wings.android.zuboradiary.data.diary.Weather;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class DiaryLiveData {
    private final MutableLiveData<LocalDate> date = new MutableLiveData<>();
    private final MutableLiveData<Weather> weather1 = new MutableLiveData<>();
    private final MutableLiveData<Weather> weather2 = new MutableLiveData<>();
    private final MutableLiveData<Condition> condition = new MutableLiveData<>();
    private final MutableLiveData<String> title = new MutableLiveData<>();
    private final MutableLiveData<Integer> numVisibleItems = new MutableLiveData<>();
    public static final int MAX_ITEMS = ItemNumber.MAX_NUMBER;
    private final DiaryItemLiveData[] items = new DiaryItemLiveData[MAX_ITEMS];
    private final MutableLiveData<Uri> picturePath = new MutableLiveData<>();
    private final MutableLiveData<LocalDateTime> log = new MutableLiveData<>();

    public DiaryLiveData() {
        for (int i = 0; i < items.length; i++) {
            int itemNumber = i + 1;
            items[i] = new DiaryItemLiveData(itemNumber);
        }
        initialize();
    }

    public void initialize() {
        date.setValue(null);
        weather1.setValue(Weather.UNKNOWN);
        weather2.setValue(Weather.UNKNOWN);
        condition.setValue(Condition.UNKNOWN);
        title.setValue("");
        numVisibleItems.setValue(1);
        for (DiaryItemLiveData item: items) {
            item.initialize();
        }
        picturePath.setValue(null);
        log.setValue(null);
    }

    public void update(DiaryEntity diaryEntity) {
        Objects.requireNonNull(diaryEntity);

        date.setValue(LocalDate.parse(diaryEntity.getDate()));
        Integer intWeather1 = getOrDefault(diaryEntity.getWeather1(), 0);
        weather1.setValue(Weather.of(intWeather1));
        Integer intWeather2 = getOrDefault(diaryEntity.getWeather2(), 0);
        weather2.setValue(Weather.of(intWeather2));
        Integer intCondition = getOrDefault(diaryEntity.getCondition(), 0);
        condition.setValue(Condition.of(intCondition));
        String title = getOrDefault(diaryEntity.getTitle(), "");
        this.title.setValue(title);

        LocalDateTime nullDateTime = null;
        String item1Title = getOrDefault(diaryEntity.getItem1Title(), "");
        String item1Comment = getOrDefault(diaryEntity.getItem1Comment(), "");
        items[0].update(item1Title, item1Comment, nullDateTime);

        String item2Title = getOrDefault(diaryEntity.getItem2Title(), "");
        String item2Comment = getOrDefault(diaryEntity.getItem2Comment(), "");
        items[1].update(item2Title, item2Comment, nullDateTime);

        String item3Title = getOrDefault(diaryEntity.getItem3Title(), "");
        String item3Comment = getOrDefault(diaryEntity.getItem3Comment(), "");
        items[2].update(item3Title, item3Comment, nullDateTime);

        String item4Title = getOrDefault(diaryEntity.getItem4Title(), "");
        String item4Comment = getOrDefault(diaryEntity.getItem4Comment(), "");
        items[3].update(item4Title, item4Comment, nullDateTime);

        String item5Title = getOrDefault(diaryEntity.getItem5Title(), "");
        String item5Comment = getOrDefault(diaryEntity.getItem5Comment(), "");
        items[4].update(item5Title, item5Comment, nullDateTime);

        int numVisibleItems = items.length;
        int maxArrayNumber  = numVisibleItems - 1;
        for (int i = maxArrayNumber; i > 0; i--) {
            if (items[i].isEmpty()) {
                numVisibleItems--;
            } else {
                break;
            }
        }
        this.numVisibleItems.setValue(numVisibleItems);

        String uriString = getOrDefault(diaryEntity.getPicturePath(), "");
        if (uriString.isEmpty()) {
            picturePath.setValue(null);
        } else {
            picturePath.setValue(Uri.parse(uriString));
        }

        log.setValue(LocalDateTime.parse(diaryEntity.getLog()));
    }

    private <T> T getOrDefault(T value, T defaultValue) {
        if (value == null) return defaultValue;
        return value;
    }

    @NonNull
    public DiaryEntity createDiaryEntity() {
        DiaryEntity diaryEntity = new DiaryEntity();
        diaryEntity.setDate(toDateString(date.getValue()));
        diaryEntity.setWeather1(toIntWeather(weather1.getValue()));
        diaryEntity.setWeather2(toIntWeather(weather2.getValue()));
        diaryEntity.setCondition(toIntCondition(condition.getValue()));
        diaryEntity.setTitle(toTrimmedString(title.getValue()));
        diaryEntity.setItem1Title(toTrimmedString(items[0].title.getValue()));
        diaryEntity.setItem1Comment(toTrimmedString(items[0].comment.getValue()));
        diaryEntity.setItem2Title(toTrimmedString(items[1].title.getValue()));
        diaryEntity.setItem2Comment(toTrimmedString(items[1].comment.getValue()));
        diaryEntity.setItem3Title(toTrimmedString(items[2].title.getValue()));
        diaryEntity.setItem3Comment(toTrimmedString(items[2].comment.getValue()));
        diaryEntity.setItem4Title(toTrimmedString(items[3].title.getValue()));
        diaryEntity.setItem4Comment(toTrimmedString(items[3].comment.getValue()));
        diaryEntity.setItem5Title(toTrimmedString(items[4].title.getValue()));
        diaryEntity.setItem5Comment(toTrimmedString(items[4].comment.getValue()));
        diaryEntity.setPicturePath(toUriString(picturePath.getValue()));
        diaryEntity.setLog(LocalDateTime.now().toString());
        return diaryEntity;
    }

    @NonNull
    private String toDateString(LocalDate date) {
        Objects.requireNonNull(date);

        return date.toString();
    }

    private int toIntWeather(Weather weather) {
        Objects.requireNonNull(weather);

        return weather.toNumber();
    }

    private int toIntCondition(Condition condition) {
        Objects.requireNonNull(condition);

        return condition.toNumber();
    }

    @NonNull
    private String toTrimmedString(String s) {
        Objects.requireNonNull(s);

        return s.trim();
    }

    @NonNull
    private String toUriString(Uri uri) {
        if (uri == null) return "";

        return uri.toString();
    }

    @NonNull
    public List<DiaryItemTitleSelectionHistoryItemEntity> createDiaryItemTitleSelectionHistoryItemEntityList() {
        List<DiaryItemTitleSelectionHistoryItemEntity> list = new ArrayList<>();
        for (int i = 0; i < DiaryLiveData.MAX_ITEMS; i++) {
            String itemTitle = items[i].title.getValue();
            LocalDateTime itemTitleUpdateLog = items[i].titleUpdateLog.getValue();
            Objects.requireNonNull(itemTitle);
            if (itemTitleUpdateLog == null) continue;
            if (itemTitle.matches("\\S+.*")) {
                DiaryItemTitleSelectionHistoryItemEntity item = new DiaryItemTitleSelectionHistoryItemEntity();
                item.setTitle(itemTitle);
                item.setLog(itemTitleUpdateLog.toString());
                list.add(item);
            }
        }
        return list;
    }

    public void incrementVisibleItemsCount() {
        Integer numVisibleItems = this.numVisibleItems.getValue();
        Objects.requireNonNull(numVisibleItems);
        Integer incrementedNumVisibleItems = numVisibleItems + 1;
        this.numVisibleItems.setValue(incrementedNumVisibleItems);
    }

    public void deleteItem(ItemNumber itemNumber) {
        getItemLiveData(itemNumber).initialize();
        Integer numVisibleItems = this.numVisibleItems.getValue();
        Objects.requireNonNull(numVisibleItems);

        if (itemNumber.getValue() < numVisibleItems) {
            for (int i = itemNumber.getValue(); i < numVisibleItems; i++) {
                ItemNumber targetItemNumber = new ItemNumber(i);
                ItemNumber nextItemNumber = new ItemNumber(i + 1);
                getItemLiveData(targetItemNumber).update(
                        getItemLiveData(nextItemNumber).title.getValue(),
                        getItemLiveData(nextItemNumber).comment.getValue(),
                        getItemLiveData(nextItemNumber).titleUpdateLog.getValue()
                );
                getItemLiveData(nextItemNumber).initialize();
            }
        }

        if (numVisibleItems > ItemNumber.MIN_NUMBER) {
            Integer decrementedNumVisibleItems = numVisibleItems - 1;
            this.numVisibleItems.setValue(decrementedNumVisibleItems);
        }
    }

    public void updateItemTitle(ItemNumber itemNumber, String title) {
        Objects.requireNonNull(title);

        getItemLiveData(itemNumber).updateItemTitle(title);
    }

    public MutableLiveData<LocalDate> getDateMutableLiveData() {
        return date;
    }

    public MutableLiveData<Weather> getWeather1MutableLiveData() {
        return weather1;
    }

    public MutableLiveData<Weather> getWeather2MutableLiveData() {
        return weather2;
    }

    public MutableLiveData<Condition> getConditionMutableLiveData() {
        return condition;
    }

    public MutableLiveData<String> getTitleMutableLiveData() {
        return title;
    }

    public MutableLiveData<Integer> getNumVisibleItemsMutableLiveData() {
        return numVisibleItems;
    }

    public DiaryItemLiveData getItemLiveData(ItemNumber itemNumber) {
        Objects.requireNonNull(itemNumber);

        int arrayNumber = itemNumber.getValue() - 1;
        return items[arrayNumber];
    }

    public MutableLiveData<Uri> getPicturePathMutableLiveData() {
        return picturePath;
    }

    public MutableLiveData<LocalDateTime> getLogMutableLiveData() {
        return log;
    }

    public static class DiaryItemLiveData {
        private final int itemNumber;
        private final MutableLiveData<String> title = new MutableLiveData<>();
        private final MutableLiveData<String> comment = new MutableLiveData<>();
        private final MutableLiveData<LocalDateTime> titleUpdateLog = new MutableLiveData<>();
        public static final int MIN_ITEM_NUMBER = 1;
        public static final int MAX_ITEM_NUMBER = 5;

        private DiaryItemLiveData(int itemNumber) {
            if (!isItemNumberInRange(itemNumber)) throw new IllegalArgumentException();

            this.itemNumber = itemNumber;
            initialize();
        }

        private boolean isItemNumberInRange(int itemNumber) {
            return itemNumber >= MIN_ITEM_NUMBER && itemNumber <= MAX_ITEM_NUMBER;
        }

        public void initialize() {
            title.setValue("");
            comment.setValue("");
            titleUpdateLog.setValue(null);
        }

        public void update(String title, String comment, @Nullable LocalDateTime titleUpdateLog) {
            Objects.requireNonNull(title);
            Objects.requireNonNull(comment);

            this.title.setValue(title);
            this.comment.setValue(comment);
            this.titleUpdateLog.setValue(titleUpdateLog);
        }

        public void updateItemTitle(String title) {
            Objects.requireNonNull(title);

            this.title.setValue(title);
            this.titleUpdateLog.setValue(LocalDateTime.now());
        }

        public boolean isEmpty() {
            String title = this.title.getValue();
            Objects.requireNonNull(title);
            String comment = this.comment.getValue();
            Objects.requireNonNull(comment);

            return title.isEmpty() && comment.isEmpty();
        }

        public int getItemNumber() {
            return itemNumber;
        }

        public MutableLiveData<String> getTitleMutableLiveData() {
            return title;
        }

        public MutableLiveData<String> getCommentMutableLiveData() {
            return comment;
        }

        public MutableLiveData<LocalDateTime> getTitleUpdateLogMutableLiveData() {
            return titleUpdateLog;
        }
    }
}

File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/diary/DiaryLiveData.kt
package com.websarva.wings.android.zuboradiary.ui.diary

import android.net.Uri
import androidx.lifecycle.MutableLiveData
import com.websarva.wings.android.zuboradiary.data.database.DiaryEntity
import com.websarva.wings.android.zuboradiary.data.database.DiaryItemTitleSelectionHistoryItemEntity
import com.websarva.wings.android.zuboradiary.data.diary.Condition
import com.websarva.wings.android.zuboradiary.data.diary.ItemNumber
import com.websarva.wings.android.zuboradiary.data.diary.Weather
import com.websarva.wings.android.zuboradiary.ui.orEmptyString
import java.time.LocalDate
import java.time.LocalDateTime

class DiaryLiveData {

    companion object {
        const val MAX_ITEMS: Int = ItemNumber.MAX_NUMBER
    }

    // MEMO:双方向DataBindingが必要の為、MutableLiveData変数はアクセス修飾子をpublicとする。
    //      LiveData変数を用意しても意味がないので作成しない。
    val date: MutableLiveData<LocalDate?> = MutableLiveData() // MEMO:初期化時日付が未定の為、null許容型とする。
    val weather1: MutableLiveData<Weather> = MutableLiveData()
    val weather2: MutableLiveData<Weather> = MutableLiveData()
    val condition: MutableLiveData<Condition> = MutableLiveData()
    val title: MutableLiveData<String> = MutableLiveData()
    val numVisibleItems: MutableLiveData<Int> = MutableLiveData()
    private val items = Array(MAX_ITEMS) { i -> DiaryItemLiveData(i + 1)}
    val picturePath: MutableLiveData<Uri?> = MutableLiveData() // MEMO:初期化時Uri有無が未定の為、null許容型とする。
    val log: MutableLiveData<LocalDateTime?> = MutableLiveData() // MEMO:初期化時日付有無が未定の為、null許容型とする。

    init {
        initialize()
    }

    fun initialize() {
        date.value = null
        weather1.value = Weather.UNKNOWN
        weather2.value = Weather.UNKNOWN
        condition.value = Condition.UNKNOWN
        title.value = ""
        numVisibleItems.value = 1
        for (item in items) {
            item.initialize()
        }
        picturePath.value = null
        log.value = null
    }

    fun update(diaryEntity: DiaryEntity) {
        date.postValue(LocalDate.parse(diaryEntity.date))
        val intWeather1 = diaryEntity.weather1
        weather1.postValue(Weather.of(intWeather1))
        val intWeather2 = diaryEntity.weather2
        weather2.postValue(Weather.of(intWeather2))
        val intCondition = diaryEntity.condition
        condition.postValue(Condition.of(intCondition))
        val title = diaryEntity.title
        this.title.postValue(title)

        val nullDateTime: LocalDateTime? = null
        val item1Title = diaryEntity.item1Title
        val item1Comment = diaryEntity.item1Comment
        items[0].update(item1Title, item1Comment, nullDateTime)

        val item2Title = diaryEntity.item2Title
        val item2Comment = diaryEntity.item2Comment
        items[1].update(item2Title, item2Comment, nullDateTime)

        val item3Title = diaryEntity.item3Title
        val item3Comment = diaryEntity.item3Comment
        items[2].update(item3Title, item3Comment, nullDateTime)

        val item4Title = diaryEntity.item4Title
        val item4Comment = diaryEntity.item4Comment
        items[3].update(item4Title, item4Comment, nullDateTime)

        val item5Title = diaryEntity.item5Title
        val item5Comment = diaryEntity.item5Comment
        items[4].update(item5Title, item5Comment, nullDateTime)

        var numVisibleItems = items.size
        val maxArrayNumber = numVisibleItems - 1
        for (i in maxArrayNumber downTo 1) {
            if (items[i].isEmpty) {
                numVisibleItems--
            } else {
                break
            }
        }
        this.numVisibleItems.postValue(numVisibleItems)

        val uriString = diaryEntity.picturePath
        if (uriString.isEmpty()) {
            picturePath.postValue(null)
        } else {
            picturePath.postValue(Uri.parse(uriString))
        }

        log.postValue(LocalDateTime.parse(diaryEntity.log))
    }

    fun createDiaryEntity(): DiaryEntity {
        return DiaryEntity(
            date.value?.toString() ?:throw IllegalStateException("日付なし(null)"),
            LocalDateTime.now().toString(),
            weather1.value?.toNumber() ?:Weather.UNKNOWN.toNumber(),
            weather2.value?.toNumber() ?:Weather.UNKNOWN.toNumber(),
            condition.value?.toNumber() ?:Condition.UNKNOWN.toNumber(),
            title.orEmptyString().trim(),
            items[0].title.orEmptyString().trim(),
            items[0].comment.orEmptyString().trim(),
            items[1].title.orEmptyString().trim(),
            items[1].comment.orEmptyString().trim(),
            items[2].title.orEmptyString().trim(),
            items[2].comment.orEmptyString().trim(),
            items[3].title.orEmptyString().trim(),
            items[3].comment.orEmptyString().trim(),
            items[4].title.orEmptyString().trim(),
            items[4].comment.orEmptyString().trim(),
            picturePath.value?.toString() ?:"",
            )
    }

    fun createDiaryItemTitleSelectionHistoryItemEntityList(): List<DiaryItemTitleSelectionHistoryItemEntity> {
        val list: MutableList<DiaryItemTitleSelectionHistoryItemEntity> = ArrayList()
        for (i in 0 until MAX_ITEMS) {
            val itemTitle = items[i].title.orEmptyString()
            val itemTitleUpdateLog = items[i].titleUpdateLog.value ?: continue
            if (itemTitle.matches("\\S+.*".toRegex())) {
                val item =
                    DiaryItemTitleSelectionHistoryItemEntity(
                        itemTitle,
                        itemTitleUpdateLog.toString()
                    )
                list.add(item)
            }
        }
        return list
    }

    fun incrementVisibleItemsCount() {
        val numVisibleItems = checkNotNull(numVisibleItems.value)
        val incrementedNumVisibleItems = numVisibleItems + 1
        this.numVisibleItems.value = incrementedNumVisibleItems
    }

    fun deleteItem(itemNumber: ItemNumber) {
        getItemLiveData(itemNumber).initialize()
        val numVisibleItems = checkNotNull(numVisibleItems.value)

        if (itemNumber.value < numVisibleItems) {
            for (i in itemNumber.value until numVisibleItems) {
                val targetItemNumber = ItemNumber(i)
                val nextItemNumber = ItemNumber(i + 1)
                getItemLiveData(targetItemNumber).update(
                    getItemLiveData(nextItemNumber).title.orEmptyString(),
                    getItemLiveData(nextItemNumber).comment.orEmptyString(),
                    getItemLiveData(nextItemNumber).titleUpdateLog.value
                )
                getItemLiveData(nextItemNumber).initialize()
            }
        }

        if (numVisibleItems > ItemNumber.MIN_NUMBER) {
            val decrementedNumVisibleItems = numVisibleItems - 1
            this.numVisibleItems.value = decrementedNumVisibleItems
        }
    }

    fun updateItemTitle(itemNumber: ItemNumber, title: String) {
        getItemLiveData(itemNumber).updateItemTitle(title)
    }

    fun getItemLiveData(itemNumber: ItemNumber): DiaryItemLiveData {
        val arrayNumber = itemNumber.value - 1
        return items[arrayNumber]
    }

    class DiaryItemLiveData(val itemNumber: Int) {

        companion object {
            const val MIN_ITEM_NUMBER: Int = ItemNumber.MIN_NUMBER
            const val MAX_ITEM_NUMBER: Int = ItemNumber.MAX_NUMBER
        }

        // MEMO:双方向DataBindingが必要の為、MutableLiveData変数はアクセス修飾子をpublicとする。
        //      LiveData変数を用意しても意味がないので作成しない。
        val title: MutableLiveData<String> = MutableLiveData()
        val comment: MutableLiveData<String> = MutableLiveData()

        // MEMO:初期化時日付有無が未定、タイトル未更新のケースがある為、null許容型とする。
        internal val titleUpdateLog: MutableLiveData<LocalDateTime?> = MutableLiveData()

        init {
            require(isItemNumberInRange(itemNumber))

            initialize()
        }

        private fun isItemNumberInRange(itemNumber: Int): Boolean {
            return itemNumber in MIN_ITEM_NUMBER..MAX_ITEM_NUMBER
        }

        fun initialize() {
            title.value = ""
            comment.value = ""
            titleUpdateLog.value = null
        }

        fun update(title: String, comment: String, titleUpdateLog: LocalDateTime?) {
            this.title.postValue(title)
            this.comment.postValue(comment)
            this.titleUpdateLog.postValue(titleUpdateLog)
        }

        fun updateItemTitle(title: String) {
            this.title.postValue(title)
            titleUpdateLog.postValue(LocalDateTime.now())
        }

        val isEmpty: Boolean
            get() {
                val title = title.orEmptyString()
                val comment = comment.orEmptyString()
                return title.isEmpty() && comment.isEmpty()
            }
    }
}
